compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/AdapterGenerator.kt

```diff
@@ -448,19 +452,38 @@ internal class AdapterGenerator(
         }
     }
 
-    /**
-     * For example,
-     * fun consumer(f: suspend () -> Unit) = ...
-     * fun nonSuspendFunction = { ... }
-     * fun useSite(...) = { ... consumer(nonSuspendFunction) ... }
-     *
-     * At the use site, instead of the argument, we can put the suspend lambda as an adapter.
-     */
-    private fun needSuspendConversion(argument: FirExpression, parameter: FirValueParameter): Boolean =
-        // TODO: should refer to LanguageVersionSettings.SuspendConversion
-        parameter.returnTypeRef.coneType.isSuspendFunctionType(session) &&
-                argument.typeRef.coneType.isBuiltinFunctionalType(session) &&
-                !argument.typeRef.coneType.isSuspendFunctionType(session)
+    private fun findInvokeSymbol(type: IrSimpleType, argument: FirExpression): IrSimpleFunctionSymbol? {
+        val argumentType = argument.typeRef.coneType
+        // Either the argument type is a non-suspend functional type
```

Comment:
Looks like these two cases might be unified: we just need to check that the value is a subtype of FunctionN, but not a SuspendFunctionN and then find invoke in its scope (it should work for both cases as far as I can see)

Reply:
I spent several hours to unify these, but two cases---a direct functional type or a subtype of functional type---always conflict to each other. :'( Maybe I'm still incorrectly finding invoke symbol, or maybe there is something more on IR lowering, but according to my trials, different symbols are required for those cases:
1) for a functional type, not sure what 'overridden' invoke symbol even means; it only works when using the declared invoke symbol. Finding it is somewhat strange too: only works when using `firSymbolProvider.getClassDeclaredCallableSymbols`, not through `scope.processFuncitonByName`.
2) for a subtype of functional type, using the overridden symbol by the current implementation or declared symbol by `scope.processFunctionByName` work.

So... I ended up splitting them, with as much intuitive function/variable names as possible.

Reply:
Answered above.

The idea is to find in the scope the `invoke` that either an override of `Function::invoke` or it is `Function::invoke` itself

Reply:
Two tests that conflict are the one I unmuted in this PR and the other one I unmuted in the previous PR
* https://github.com/JetBrains/kotlin/blob/master/compiler/testData/codegen/box/coroutines/suspendConversion/onArgument.kt
* https://github.com/JetBrains/kotlin/blob/master/compiler/testData/codegen/box/coroutines/suspendConversion/subtypeOfFunctionalTypeToSuspendConversion.kt

For `onArgument`, neither an override of `Function::invoke` or `Function::invoke` in the scope of argument type (`Function1<String, String>`) works. It only works when the type-unsubstituted, base `invoke` symbol (`Function1<P1, R>`) is used. Note that the base invoke symbol won't work for the other use-site: type constraint building for arguments, because the type argument (`String`) is needed as constraints.